refactor: frontend of `trace_decoder`

0xaatif opened the pull request:
- Part of #275 

The previous code was _extremely_ difficult to reason about, and was blocking prior attempts at the above issue.
The best thing was a rewrite, followed by SMT (backend) support (to be completed in a later PR).

For an explanation of the terminology, see this:
https://github.com/0xPolygonZero/zk_evm/blob/76669ee601331bde8b6e516643baa854081fdeea/trace_decoder/src/lib.rs#L68-L78

# Change summary
- Changes the public API of `trace_decoder` to be a single `fn entrypoint`, and a bunch of structs at the root.
  I'd prefer this PR didn't bikeshed names - we can't finalize an API until we see what the SMT backend requires.
- Rewrite the binary parsing in `wire`.
- Rewrite the binary frontend in `zero_jerigon`
- Ports the old MPT test cases, giving us a high degree of confidence in the changes to the frontend.
- Introduces (test only) frontend for TBC SMT support in `hermez_cdk_erigon`
- This PR also changed structs that are deserialized from RPC nodes, so @temaniarpit27 did live testing, report [here](https://www.notion.so/polygontechnology/Trace-Decoder-Testing-e19f240155864d0eaa34c3f03d43ecd4)

Nashtare commented:
@temaniarpit27 As you are in charge of testing, could we have (in addition to whole, regular blocks), tests for these "edge-cases":

- an empty block
- an empty block with withdrawals
- a block containing a single transaction
- a block containing a single transaction with withdrawals

0xaatif commented:
> * testing: the previous decoder was testing some specific test payloads, which this PR completely removes (including doctest). Current coverage is even lower than before (37.8% -> 29.8% for `trace_decoder` only) so may deserve more work

@Nashtare could you help me understand your objection here?
AIUI the library contained no doctests previously, only unnecessary code snippets, where I've moved comments to the appropriate places in code, or deleted, which I think is the right thing to do. (why paste a copy of e.g the zero-bin callsite, to just get out of date??)
```console
$ git rev-parse --short HEAD
baecefcb
$ cargo test --package trace_decoder --doc
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.11s
   Doc-tests trace_decoder

running 3 tests
test trace_decoder/src/lib.rs - (line 24) ... ignored
test trace_decoder/src/lib.rs - (line 6) ... ignored
test trace_decoder/src/lib.rs - (line 76) ... ignored
```

It contained 6 MPT payloads:
https://github.com/0xPolygonZero/zk_evm/blob/baecefcbaa411c77fb5135c626a954c7d88f3028/trace_decoder/src/compact/complex_test_payloads.rs#L17-L30

All of which I have migrated:
https://github.com/0xPolygonZero/zk_evm/blob/76669ee601331bde8b6e516643baa854081fdeea/trace_decoder/src/test_cases/zero_jerigon.json

There is only one test case that has been removed:
https://github.com/0xPolygonZero/zk_evm/blob/baecefcbaa411c77fb5135c626a954c7d88f3028/trace_decoder/src/compact/compact_prestate_processing.rs#L1448

Which I don't trust because of this:
https://github.com/0xPolygonZero/zk_evm/blob/76669ee601331bde8b6e516643baa854081fdeea/trace_decoder/src/wire.rs#L241-L244

The other point you mentioned is coverage - code coverage of the backend should be the same, because that's not changed,
Could you point me to the coverage drop for the frontend? Can it not be attributed to the drop in code size?

Nashtare commented:
My bad for doctest, I thought these were running and not ignored.
As for coverage, this is what I got for `trace_decoder`:

`develop`:

> || trace_decoder/src/compact/compact_prestate_processing.rs: 334/488
|| trace_decoder/src/compact/compact_to_partial_trie.rs: 66/82
|| trace_decoder/src/decoding.rs: 0/322
|| trace_decoder/src/deserializers.rs: 0/23
|| trace_decoder/src/processed_block_trace.rs: 0/158
|| trace_decoder/src/trace_protocol.rs: 0/4
|| trace_decoder/src/utils.rs: 15/22

`this_branch`:

> || trace_decoder/src/decoding.rs: 0/327
|| trace_decoder/src/lib.rs: 8/70
|| trace_decoder/src/processed_block_trace.rs: 0/82
|| trace_decoder/src/wire.rs: 81/101
|| trace_decoder/src/zero_jerigon.rs: 137/179

0xaatif commented:
@Nashtare 
I think we agree that:
- WIBNI to get coverage up, and perhaps @temaniarpit27's testing will

But I hope I've shown that, since the test cases are the same the drop in %coverage is an apples to oranges comparison, so shouldn't block this review.
In fact, it's not surprising, since the previous code had many trivial indirections which would have bloated covered line count for the same logical path.

BGluth commented on trace_decoder/src/decoding.rs:
```diff
@@ -149,27 +154,19 @@ pub enum TraceParsingErrorReason {
     /// Failure due to trying to withdraw from a missing account
     #[error("No account present at {0:x} (hashed: {1:x}) to withdraw {2} Gwei from!")]
-    MissingWithdrawalAccount(Address, HashedAccountAddr, U256),
+    MissingWithdrawalAccount(Address, H256, U256),
```
Yeah this is probably a bit subjective, but I agree with Robin here in that I think using aliases to make it clear what the data actually is here makes signatures more readable.

BGluth commented:
@0xaatif For the next step of refactoring the delta application logic to work with both MPT & SMT, are you thinking about going the route of abstracting away trie operations between both trie types?

0xaatif commented:
> @0xaatif For the next step of refactoring the delta application logic to work with both MPT & SMT, are you thinking about going the route of abstracting away trie operations between both trie types?

@BGluth let's discuss this in #275 or in Slack

0xaatif commented on trace_decoder/src/decoding.rs:
```diff
@@ -83,7 +88,7 @@ impl std::error::Error for TraceParsingError {}
 impl TraceParsingError {
     /// Function to create a new TraceParsingError with mandatory fields
-    pub(crate) fn new(reason: TraceParsingErrorReason) -> Self {
+    fn new(reason: TraceParsingErrorReason) -> Self {
```
I'll defer this to when I refactor the backend.

0xaatif commented on trace_decoder/src/hermez_cdk_erigon.rs:
```diff
@@ -0,0 +1,244 @@
+
+/// # Panics
+/// - Liberally, both in this module and the [`smt_trie`] library. Therefore, do
+///   NOT call this function on untrusted inputs
```
I would prefer to defer this until `smt_trie` has documentation on when it panics.

0xaatif commented on trace_decoder/src/decoding.rs:
```diff
@@ -305,7 +294,7 @@ impl ProcessedBlockTrace {
         trie_state: &mut PartialTrieState,
         meta: &TxnMetaState,
-        txn_idx: TxnIdx,
+        txn_idx: usize,
```
It's my preference.

I don't think they're particularly useful when used like this - I prefer actually naming variables.
The above line is actually _more_ confusing with the type alias - is there a special `TxnIdx` struct?
No, it's just a `#define usize`, for no particular reason...

@BGluth let's have this discussion in one place

BGluth replied:
Yeah fair enough. You're taking over the refactor, so you can have the creative license. :slightly_smiling_face:

While I agree with you about preferring indicating the usage of a variable on the variable name (since it means an extra indirection to read the type), there are a few contexts where we can't or it's hard to specify nicely (eg. is the enum tuple variant expect a hashed slot or just a slot, or is `storage_accesses: Vec<(H256, StorageAccess)>,` using hashed or unhashed slots for keys?).

0xaatif replied:
Those make sense, I particularly think that e.g having `fn(&CodeHash) -> Vec<u8>` gets the point across. I might re-introduce them, or create newtypes as makes sense :)

0xaatif commented:
> @temaniarpit27 As you are in charge of testing, could we have (in addition to whole, regular blocks), tests for these "edge-cases"

@Nashtare, I think this PR has the following main risks:
1. Breaking deserialization, because in the `BlockData`/`OtherBlockData` tree, I've removed some indirections, enum variants, and changed deserializers.
2. Breaking frontend transformations.

(2) is partially addressed by the existing unit tests, which we've already discussed.
I've asked Arpit to do integration testing to further de-risk the above.

I welcome additional test cases, but they shouldn't block this `refactor:` PR, and in turn progress on the SMT work - they belong to a future `test:` PR.

Do you agree?

Nashtare commented on trace_decoder/src/wire.rs:
```diff
@@ -0,0 +1,314 @@
+        key: NonEmpty<Vec<U4>>,
+        value: NonEmpty<Vec<u8>>,
+    },
+    // BUG: the spec has `EXTENSION` and `ACCOUNT_LEAF` inconsistent
+    // between the instruction list and encoding list
```
which `spec` is buggy here, the one referenced for type-1 or type-2?

Nashtare commented on trace_decoder/src/wire.rs:
```diff
@@ -0,0 +1,314 @@
+            trace("code", cbor.map(|raw_code| Instruction::Code { raw_code })).parse_next(input)
+        }
+        0x05 => trace("account_leaf", account_leaf).parse_next(input),
+        // BUG: this opcode is is undocumented, but the previous version of
+        //      this code had it, and our tests fail without it
```
```suggestion
        // BUG: this opcode is undocumented, but the previous version of
        //      this code had it, and our tests fail without it.
```

Nashtare replied:
Also may be worth opening a tracking issue for it.

0xaatif replied:
#349

Nashtare commented on trace_decoder/src/zero_jerigon.rs:
```diff
@@ -0,0 +1,456 @@
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum Node {
+    Hash(Hash),
+    // BUG: these are documented, but never constructed during execution
+    // Value(Value),
+    // Account(Account),
```
I assume you'll remove them when dealing with the backend then?

0xaatif replied:
I think this should wait for #349

Nashtare commented on trace_decoder/src/lib.rs:
```diff
@@ -109,30 +63,395 @@
-#![feature(trait_alias)]
-#![feature(iter_array_chunks)]
 #![deny(rustdoc::broken_intra_doc_links)]
-#![deny(missing_debug_implementations)]
-#![deny(missing_docs)]
```
These are good to have IMHO, would deserve a TODO to not get forgotten and possibly addressed later.

0xaatif replied:
restored to `warn`, they'll fail in CI because we `-D warnings`, but shouldn't hamper local development

muursh commented on trace_decoder/src/hermez_cdk_erigon.rs:
```diff
@@ -0,0 +1,244 @@
+//! Frontend for the witness format emitted by [`0xPolygonHermez/cdk-erigon`](https://github.com/0xPolygonHermez/cdk-erigon/)
```
I'm not sure we should name the file this. The repo might actually be moved soon so it wouldn't be `hermez...` I think a more generic name is better but that's also a very fussy opinion.

0xaatif replied:
I've named the files this way because they reflect vendor-specific support. I'm not fussed by the actual name, but "type1" and "type2" didn't really sit right. Happy to take suggestions

muursh replied:
`type 1` and `type 2` is the standard nomenclature. It's not just "ah this is v1 so let's call this v2". See [here](https://vitalik.eth.limo/general/2022/08/04/zkevm.html).

muursh replied:
> I understand, but I think the type2 => hermez and type1 => plonky are polygon-specific aliases.

plonky2 will be both t1 and t2, I mean it already is but t2 is still in a different branch for a little bit longer. Hermez is our existing t2 prover which means we tend to conflate them but strictly speaking it's incorrect to do so. 

> My worry was that calling this module type1 implied that the logic would work for all type1 provers

All of this code will only work with plonky2 not with hermez inc. the t2 stuff.

I vote going type1 and type2.

muursh commented on trace_decoder/src/hermez_cdk_erigon.rs:
```diff
@@ -0,0 +1,244 @@
+    pub collation: HashMap<ethereum_types::Address, CollatedLeaf>,
+}
+
+/// # Panics
+/// - Liberally, both in this module and the [`smt_trie`] library. Therefore, do
+///   NOT call this function on untrusted inputs.
+pub fn frontend(instructions: impl IntoIterator<Item = Instruction>) -> anyhow::Result<Frontend> {
```
Long term I'd rather move away from panicking as much as possible but we can deal with that another day

0xaatif commented on trace_decoder/src/processed_block_trace.rs:
```diff
@@ -4,272 +4,58 @@ use std::iter::once;
-struct CodeHashResolving<F> {
+pub(crate) struct CodeHashResolving<F> {
     /// If we have not seen this code hash before, use the resolve function that
     /// the client passes down to us. This will likely be an rpc call/cache
     /// check.
-    client_code_hash_resolve_f: F,
+    pub client_code_hash_resolve_f: F,
 
     /// Code hash mappings that we have constructed from parsing the block
     /// trace. If there are any txns that create contracts, then they will also
     /// get added here as we process the deltas.
-    extra_code_hash_mappings: HashMap<CodeHash, Vec<u8>>,
+    pub extra_code_hash_mappings: HashMap<H256, Vec<u8>>,
 }
```
The struct is `pub(crate)` because I don't want it it leaking. The fields are `pub` because I need to set the from the parent module.

0xaatif commented on trace_decoder/benches/block_processing.rs:
```diff
@@ -5,37 +5,27 @@
+use trace_decoder::{BlockTrace, OtherBlockData};
 
-#[derive(Clone, Debug, Deserialize, Serialize)]
+#[derive(Clone, Debug, serde::Deserialize)]
```
```rust
use serde::Deserialize;

#[derive(Deserialize)
struct Foo;
```

```rust
#[derive(serde::Deserialize)]
```

We don't use the trait/macro anywhere else in this file.
struct Foo;
```

0xaatif commented on trace_decoder/src/decoding.rs:
```diff
@@ -21,22 +21,27 @@ use mpt_trie::{
 /// failure.
 pub type TraceParsingResult<T> = Result<T, Box<TraceParsingError>>;
 
+const EMPTY_ACCOUNT_BYTES_RLPED: [u8; 70] = [
```
This is moved from a deleted file - I also want to see these constants deleted, and I have where possible

0xaatif commented on trace_decoder/src/wire.rs:
```diff
@@ -0,0 +1,314 @@
+fn instruction(input: &mut &[u8]) -> PResult<Instruction> {
+    let start = input.checkpoint();
+    let opcode = any(input)?;
+    // I don't like the `winnow::combinator::dispatch!` macro
```
There's no macro here. Clarified

0xaatif commented on trace_decoder/src/hermez_cdk_erigon.rs:
```diff
@@ -0,0 +1,244 @@
+            other => Some(other),
+        });
+    let folded = fold1(&mut instructions)?.context("no instructions to fold")?;
+    // this will allow trailing Code instructions, which isn't as strict as we
```
changed wording

0xaatif commented:
yikes @ CI, fixed in #357

muursh commented:
I think I'm broadly good with this

atanmarko commented on trace_decoder/src/lib.rs:
```diff
@@ -109,30 +63,402 @@
+///
+/// Deviations from the specification are signalled with `BUG(spec)` in the
+/// code.
+const _DEVELOPER_DOCS: () = ();
```
Do you still need this?

0xaatif replied:
normally developer documentation is `// <some docs>` but they run the risk of getting stale that way if they link to items in the crate. This is the way I do developer docs with intra-doc links. I _do_ think this developer documentation is valuable, so I want to keep it

atanmarko commented on trace_decoder/src/lib.rs:
```diff
@@ -109,30 +63,402 @@
+use mpt_trie::partial_trie::HashedPartialTrie;
+use serde::{Deserialize, Serialize};
+
+/// Core payload needed to generate a proof for a block. Note that the scheduler
+/// may need to request some additional data from the client along with this in
+/// order to generate a proof.
```
I don't know if this is the topic for this PR but I find the old terminology is quite confusing (scheduler, client)

```suggestion
/// Core payload needed to generate proof for a block. Additional data retrievable from the blockchain node (using standard ETH RPC API) may be needed for proof generation.
```

0xaatif replied:
I barely know what the domain objects are here - thanks for helping to clear up the docs :)

atanmarko commented on trace_decoder/src/lib.rs:
```diff
@@ -109,30 +63,402 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub nonce: Option<U256>,
+
+    /// Account addresses that were only read by the txn.
```
```suggestion
    /// Account storage locations that were only read by the txn.
```

atanmarko commented on trace_decoder/src/lib.rs:
```diff
@@ -109,30 +63,402 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub storage_read: Option<Vec<H256>>,
+
+    /// Account storage addresses that were mutated by the txn along with their
+    /// new value.
```
```suggestion
    /// Account storage locations that were mutated by the txn along with their
    /// new value.
```

atanmarko commented on trace_decoder/src/lib.rs:
```diff
@@ -109,30 +63,402 @@
+
+/// TODO(0xaatif): https://github.com/0xPolygonZero/zk_evm/issues/275
+///                document this once we have the API finalized
+pub fn entrypoint(
```
Non-blocking: As this will be the only API function of the trace decoder I wonder if some other name would be more appropriate, e.g. `process`, or `generate_inputs`?

0xaatif replied:
> I'd prefer this PR didn't bikeshed names - we can't finalize an API until we see what the SMT backend requires.

From the review description :)

atanmarko commented on trace_decoder/src/lib.rs:
```diff
@@ -109,30 +63,402 @@
+    use crate::processed_block_trace::{
+        CodeHashResolving, ProcessedBlockTrace, ProcessedBlockTracePreImages,
+    };
+    use crate::PartialTriePreImages;
+    use crate::{
+        BlockTraceTriePreImages, CombinedPreImages, SeparateStorageTriesPreImage,
+        SeparateTriePreImage, SeparateTriePreImages,
+    };
```
Suggestion:
```suggestion
    use crate::{
        BlockTraceTriePreImages, CombinedPreImages, SeparateStorageTriesPreImage,
        SeparateTriePreImage, SeparateTriePreImages, PartialTriePreImages
    };
```

0xaatif replied:
This block is removed in backend support :)

atanmarko commented on trace_decoder/src/processed_block_trace.rs:
```diff
@@ -4,272 +4,58 @@ use std::iter::once;
+    fn resolve(&mut self, c_hash: &H256) -> Vec<u8> {
         match self.extra_code_hash_mappings.get(c_hash) {
             Some(code) => code.clone(),
             None => (self.client_code_hash_resolve_f)(c_hash),
```
Non-blocking: Would we ever really need to dynamically retrieve the hash or we could always get it before the trace decoder starts working? We would probably have to spin up tokio async engine instance and combine multiple resolve calls with this one engine

BGluth replied:
We could, but we would need to do a separate pass over the block trace and all of the compact, which I think will be a bit slower (but it's already pretty fast anyways I guess). Right now, this resolve function is never actually hit. We always have the matching code for all code hashes embedded inside of the compact encoding pre-images (with the exception of when contract creation occurs).

0xaatif replied:
As discussed in slack, my short-term goal is to remove this. we can add it back when it's needed.

atanmarko commented on trace_decoder/src/wire.rs:
```diff
@@ -0,0 +1,317 @@
+    let actual = parser.parse(src.as_ref()).unwrap();
+    assert_eq!(expected, actual)
+}
+
```
Non blocking: should we add extensive testing here to cover [specification](https://github.com/0xPolygonHermez/cdk-erigon/blob/zkevm/turbo/trie/WITNESS.md)

0xaatif replied:
#362

0xaatif commented:
@muursh / @Nashtare what will it take to get this merged? I'm starting to get bogged down by keeping up with `develop`

muursh commented:
I looked at an earlier version and had no major concerns. If Robin has looked and is happy then I'd say you're free to merge.

Nashtare reviewed (approved):
Yeah I'm broadly ok to merge (based on previous version I reviewed) + latest review by Marko.
